hal-x86_64/src/log.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{serial, vga};
+use log::{Level, LevelFilter, Log, Metadata, Record};
+
+// #[derive(Debug)]
+pub struct Logger {
+    vga_max_level: LevelFilter,
+    serial: Option<Serial>,
+}
+
+struct Serial {
+    port: &'static serial::Port,
+    max_level: LevelFilter,
+}
+
+impl Default for Logger {
+    fn default() -> Self {
+        Self {
+            vga_max_level: LevelFilter::Trace,
+            serial: serial::com1().map(|port| Serial {
+                port,
+                max_level: LevelFilter::Trace,
+            }),
+        }
+    }
+}
+
+impl Logger {
+    pub const fn vga_only(vga_max_level: LevelFilter) -> Self {
+        Self {
+            vga_max_level,
+            serial: None,
+        }
+    }
+
+    #[inline]
+    fn vga_enabled(&self, level: Level) -> bool {
+        level <= self.vga_max_level
+    }
+
+    #[inline]
+    fn serial_enabled(&self, level: Level) -> bool {
+        self.serial
+            .as_ref()
+            .map(|serial| level <= serial.max_level)
+            .unwrap_or(false)
+    }
+}
+
+impl Log for Logger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        let level = metadata.level();
+        self.vga_enabled(level) || self.serial_enabled(level)
+    }
+
+    fn log(&self, record: &Record) {
+        let level = record.level();
+        if self.vga_enabled(level) {
+            use core::fmt::Write;
+            let mut vga = vga::writer();
```

Comment:
In the case of these special types which are intended to be used with `write!`-style macros, we may want to add an inherent `write_fmt` method, like what `core` does with `Formatter`. It allows you to avoid importing the `Write` trait to write into the type. https://github.com/rust-lang/rust/blob/0fb43801368ae8b5931583f813071120bed55c35/src/libcore/fmt/mod.rs#L1468-L1489